internal/mode/static/state/graph/route_common.go

```diff
@@ -529,6 +534,30 @@ func routeAllowedByListener(
 	return true
 }
 
+// isRouteKindAllowedByListener checks if the route is allowed to attach to the listener.
+// If the listener specifies allowed kinds, the route kind must be in the list.
+// If the listener does not specify allowedRoutes, allowed routes are determined using the listener protocol.
+func isRouteTypeAllowedByListener(listener *Listener, routeType RouteType) bool {
+	if listener.Source.AllowedRoutes != nil && listener.Source.AllowedRoutes.Kinds != nil {
+		for _, kind := range listener.Source.AllowedRoutes.Kinds {
+			switch kind.Kind {
+			case kinds.HTTPRoute:
+				return routeType == RouteTypeHTTP
+			case kinds.GRPCRoute:
+				return routeType == RouteTypeGRPC
+			}
+		}
```

Comment:
This will fail if the routeType is GRPCRoute and the listener has kinds HTTPRoute and GRPCRoute specified, because the first condition will return false.

We'll a test case for this as well. Listener has both HTTPRoute and GRPCRoute kinds allowed, so both should work.